packages/multi-select-combo-box/src/vaadin-multi-select-combo-box-internal.js

```diff
@@ -246,6 +269,54 @@ class MultiSelectComboBoxInternal extends ComboBoxDataProviderMixin(ComboBoxMixi
       );
     }
   }
+
+  /**
+   * Override method inherited from the combo-box
+   * to render only selected items when read-only,
+   * even if a different set of items is provided.
+   *
+   * @protected
+   * @override
+   */
+  _setOverlayItems(items) {
+    const effectiveItems = this.readonly ? this.selectedItems : items;
+    super._setOverlayItems(effectiveItems);
+  }
+
+  /**
+   * Override method inherited from the combo-box
+   * to not request data provider when read-only.
+   *
+   * @param {number}
+   * @return {boolean}
+   * @protected
+   * @override
+   */
+  _shouldLoadPage(page) {
+    if (this.readonly) {
+      return false;
+    }
+
+    return super._shouldLoadPage(page);
+  }
+
+  /** @private */
+  _readonlyChanged(readonly, oldReadonly) {
+    if (readonly) {
+      this.__savedItems = this._getOverlayItems();
+      this._setOverlayItems(this.selectedItems);
+    } else if (oldReadonly) {
+      this._setOverlayItems(this.__savedItems);
+      this.__savedItems = null;
+    }
```

Comment:
Moved this observer from the host element to avoid accessing protected methods from outside.

Reply:
This PR has 2 separate observers for `readonly` that both do pretty similar-looking stuff (`this._setOverlayItems`). 

Should we merge them under one multi-property observer to avoid
- a possible race condition
- accessing `this.selectedItems` without it actually being a dependency of this observer
- having to override `readonly` in properties just to add an observer

One downside would be that the `oldReadOnly` would need to be maintained manually (a convenience you get with simple observers)?

Reply:
Thanks, updated to use multi-property observer with `oldReadOnly` flag maintained manually.